tests/test_catalog.py

```diff
@@ -837,6 +837,27 @@ def check_all_absolute(cat: Catalog) -> None:
                 c2.catalog_type = CatalogType.ABSOLUTE_PUBLISHED
                 check_all_absolute(c2)
 
+    def test_self_contained_catalog_collection_item_links(self) -> None:
+        with tempfile.TemporaryDirectory() as tmp_dir:
+            cat = TestCases.test_case_9()
+            cat.normalize_hrefs(tmp_dir)
+            cat.validate_all()
+            print(f"Saving catalog to: {tmp_dir}")
+            cat.save(catalog_type=CatalogType.SELF_CONTAINED)
+
+            # read the item back, to make experiment clean
+            item = Item.from_file(
+                f"{tmp_dir}/collection-issue-657/item-issue-657/item-issue-657.json"
+            )
+            # ensure that all links in the output json are relative
+            # everything of a nested level > 2 is not relative?
+            for link in item.links:
+                href = link.get_href()
+                self.assertIsNotNone(href)
+                self.assertFalse(is_absolute_href(href))
```

Comment:
Is there a way around the following error: 

```python
tests/test_catalog.py:858: error: Argument 1 to "is_absolute_href" has incompatible type "Optional[str]"; expected "str"  [arg-type]
Found 1 error in 1 file (checked 88 source files)
```

Reply:
Sorry, I led you a little astray there. I thought `mypy` would recognize `self.assertIsNotNone` as eliminating the possibility of a `None` type for that variable, but it looks like it does not. 

Using `assert href is not None` instead of `self.assertIsNotNone(href)` fixes that issue.